src/compiler/types.ts

```diff
@@ -2343,6 +2345,13 @@ namespace ts {
     // see: https://tc39.github.io/ecma262/#prod-SuperProperty
     export type SuperProperty = SuperPropertyAccessExpression | SuperElementAccessExpression;
 
+    export interface PrivateIdentifierInInExpression extends Expression {

```

Comment:
It seems intentional to solve the issue like this, so this comment is just a question.
Although it is pretty clear to add another expression like this to handle the situation, however, Why not just enhance existing "in" narrow function `narrowByInKeyword`?

Reply:
Having the new expression type is more for the parsing and the emitter than the type check. `in` is a `BinaryExpression` but the LHS here is not an expression so it would break the assumptions of all the existing code around BinaryExpressions.